// ionic-framework/e2e/tabs/basic/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var ionic_1 = require('ionic/ionic');
//
// Tab 1
//
var Tab1 = (function () {
    function Tab1(nav) {
        this.nav = nav;
        this.items = [];
        for (var i = 1; i <= 250; i++) {
            this.items.push(i);
        }
    }
    Tab1 = __decorate([
        ionic_1.Page({
            template: "\n    <ion-navbar *navbar>\n      <ion-title>Heart</ion-title>\n    </ion-navbar>\n    <ion-content>\n      <ion-list>\n        <ion-list-header>\n          Tab 1\n        </ion-list-header>\n        <ion-item *ng-for=\"#i of items\">Item {{i}} {{i}} {{i}} {{i}}</ion-item>\n      </ion-list>\n    </ion-content>\n    "
        }), 
        __metadata('design:paramtypes', [(typeof (_a = typeof ionic_1.NavController !== 'undefined' && ionic_1.NavController) === 'function' && _a) || Object])
    ], Tab1);
    return Tab1;
    var _a;
})();
//
// Tab 2
//
var Tab2 = (function () {
    function Tab2() {
        this.sessions = [];
        for (var i = 1; i <= 250; i++) {
            this.sessions.push({
                name: 'Name ' + i,
                location: 'Location: ' + i
            });
        }
    }
    Tab2 = __decorate([
        ionic_1.Page({
            template: "\n    <ion-navbar *navbar>\n      <ion-title>Schedule</ion-title>\n    </ion-navbar>\n    <ion-content>\n      <ion-list>\n        <ion-item-sliding *ng-for=\"#session of sessions\" #sliding-item>\n          <ion-item>\n            <h3>{{session.name}} {{session.name}} {{session.name}}</h3>\n            <p>{{session.location}} {{session.location}} {{session.location}}</p>\n          </ion-item>\n          <ion-item-options>\n            <button primary>Speaker<br>Info</button>\n            <button secondary>Add to<br>Favorites</button>\n          </ion-item-options>\n        </ion-item-sliding>\n      </ion-list>\n    </ion-content>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], Tab2);
    return Tab2;
})();
//
// Tab 3
//
var Tab3 = (function () {
    function Tab3() {
    }
    Tab3 = __decorate([
        ionic_1.Page({
            template: "\n    <ion-navbar *navbar>\n      <button menu-toggle>\n        <icon menu></icon>\n      </button>\n      <ion-title>Stopwatch</ion-title>\n    </ion-navbar>\n    <ion-content padding>\n      <h2>Tab 3</h2>\n    </ion-content>\n    "
        }), 
        __metadata('design:paramtypes', [])
    ], Tab3);
    return Tab3;
})();
var TabsPage = (function () {
    function TabsPage() {
        this.root1 = Tab1;
        this.root2 = Tab2;
        this.root3 = Tab3;
    }
    TabsPage = __decorate([
        ionic_1.App({
            template: "\n    <ion-menu [content]=\"content\">\n      <ion-toolbar secondary>\n        <ion-title>Menu</ion-title>\n      </ion-toolbar>\n      <ion-content>\n        <ion-list>\n          <button ion-item menu-close detail-none>\n            Close Menu\n          </button>\n        </ion-list>\n      </ion-content>\n    </ion-menu>\n\n    <ion-tabs #content>\n      <ion-tab tab-title=\"Plain List\" tab-icon=\"star\" [root]=\"root1\"></ion-tab>\n      <ion-tab tab-title=\"Schedule\" tab-icon=\"globe\" [root]=\"root2\"></ion-tab>\n      <ion-tab tab-title=\"Stopwatch\" tab-icon=\"stopwatch\" [root]=\"root3\"></ion-tab>\n    </ion-tabs>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], TabsPage);
    return TabsPage;
})();
exports.TabsPage = TabsPage;
