// awesome-typescript-loader/dist.babel/instance.js
"use strict";

var __awaiter = undefined && undefined.__awaiter || function (thisArg, _arguments, Promise, generator) {
    return new Promise(function (resolve, reject) {
        generator = generator.call(thisArg, _arguments);
        function cast(value) {
            return value instanceof Promise && value.constructor === Promise ? value : new Promise(function (resolve) {
                resolve(value);
            });
        }
        function onfulfill(value) {
            try {
                step("next", value);
            } catch (e) {
                reject(e);
            }
        }
        function onreject(value) {
            try {
                step("throw", value);
            } catch (e) {
                reject(e);
            }
        }
        function step(verb, value) {
            var result = generator[verb](value);
            result.done ? resolve(result.value) : cast(result.value).then(onfulfill, onreject);
        }
        step("next", void 0);
    });
};
var host_1 = require('./host');
var colors = require('colors');
var fs = require('fs');
var path = require('path');
var _ = require('lodash');
var tsconfig = require('tsconfig');
var helpers_1 = require('./helpers');
var deps_1 = require('./deps');
var checker_1 = require('./checker');
var tsconfig_utils_1 = require('./tsconfig-utils');
var resolver_1 = require('./resolver');
var pkg = require('../package.json');
function getRootCompiler(compiler) {
    if (compiler.parentCompilation) {
        return getRootCompiler(compiler.parentCompilation.compiler);
    } else {
        return compiler;
    }
}
function getInstanceStore(compiler) {
    var store = getRootCompiler(compiler)._tsInstances;
    if (store) {
        return store;
    } else {
        throw new Error('Can not resolve instance store');
    }
}
function ensureInstanceStore(compiler) {
    var rootCompiler = getRootCompiler(compiler);
    if (!rootCompiler._tsInstances) {
        rootCompiler._tsInstances = {};
    }
}
function resolveInstance(compiler, instanceName) {
    return getInstanceStore(compiler)[instanceName];
}
var COMPILER_ERROR = colors.red("\n\nTypescript compiler cannot be found, please add it to your package.json file:\n    npm install --save-dev typescript\n");
var BABEL_ERROR = colors.red("\n\nBabel compiler cannot be found, please add it to your package.json file:\n    npm install --save-dev babel\n");
function ensureInstance(webpack, options, instanceName) {
    ensureInstanceStore(webpack._compiler);
    var exInstance = resolveInstance(webpack._compiler, instanceName);
    if (exInstance) {
        return exInstance;
    }
    var tsFlow = Promise.resolve();
    var compilerName = options.compiler || 'typescript';
    var compilerPath = path.dirname(compilerName);
    if (compilerPath == '.') {
        compilerPath = compilerName;
    }
    var tsImpl = undefined;
    try {
        tsImpl = require(compilerPath);
    } catch (e) {
        console.error(e);
        console.error(COMPILER_ERROR);
        process.exit(1);
    }
    var libPath = path.join(compilerPath, 'lib', 'lib.d.ts');
    var lib6Path = path.join(compilerPath, 'lib', 'lib.es6.d.ts');
    try {
        require.resolve(libPath);
    } catch (e) {
        libPath = path.join(compilerPath, 'bin', 'lib.d.ts');
        lib6Path = path.join(compilerPath, 'bin', 'lib.es6.d.ts');
    }
    var compilerInfo = {
        compilerName: compilerName,
        compilerPath: compilerPath,
        tsImpl: tsImpl,
        lib5: helpers_1.loadLib(libPath),
        lib6: helpers_1.loadLib(lib6Path)
    };
    _.defaults(options, {
        resolveGlobs: true
    });
    var configFilePath = undefined;
    var configFile = undefined;
    var folder = options.tsconfig || process.cwd();
    configFilePath = tsconfig.resolveSync(folder);
    if (configFilePath) {
        configFile = tsconfig_utils_1.parseContent(fs.readFileSync(configFilePath).toString(), configFilePath);
        if (options.resolveGlobs) {
            tsconfig_utils_1.tsconfigSuggestions(configFile);
            configFile = tsconfig.readFileSync(configFilePath);
        }
    }
    var tsFiles = [];
    if (configFile) {
        if (configFile.compilerOptions) {
            _.extend(options, configFile.compilerOptions);
            _.extend(options, configFile.awesomeTypescriptLoaderOptions);
            tsFiles = configFile.files;
        }
    }
    options = tsconfig_utils_1.rawToTsCompilerOptions(options, path.dirname(configFilePath), tsImpl);
    _.defaults(options, {
        externals: [],
        doTypeCheck: true,
        sourceMap: true,
        verbose: false,
        noLib: false,
        sourceRoot: process.cwd()
    });
    options = _.omit(options, 'outDir', 'files', 'exclude');
    options.externals.push.apply(options.externals, tsFiles);
    var babelImpl = undefined;
    if (options.useBabel) {
        try {
            babelImpl = require(path.join(process.cwd(), 'node_modules', 'babel-core'));
        } catch (e) {
            console.error(BABEL_ERROR);
            process.exit(1);
        }
    }
    var cacheIdentifier = null;
    if (options.useCache) {
        if (!options.cacheDirectory) {
            options.cacheDirectory = path.join(process.cwd(), '.awcache');
        }
        if (!fs.existsSync(options.cacheDirectory)) {
            fs.mkdirSync(options.cacheDirectory);
        }
        cacheIdentifier = {
            'typescript': tsImpl.version,
            'awesome-typescript-loader': pkg.version,
            'awesome-typescript-loader-query': webpack.query,
            'babel-core': babelImpl ? babelImpl.version : null
        };
    }
    var forkChecker = options.forkChecker && getRootCompiler(webpack._compiler)._tsFork;
    var resolver = resolver_1.default(webpack._compiler.options);
    var syncResolver = resolver.resolveSync.bind(resolver);
    var tsState = new host_1.State(options, webpack._compiler.inputFileSystem, compilerInfo, syncResolver);
    var compiler = webpack._compiler;
    setupWatchRun(compiler, instanceName);
    if (options.doTypeCheck) {
        setupAfterCompile(compiler, instanceName, forkChecker);
    }
    var webpackOptions = _.pick(webpack._compiler.options, 'resolve');
    return getInstanceStore(webpack._compiler)[instanceName] = {
        tsFlow: tsFlow,
        tsState: tsState,
        babelImpl: babelImpl,
        compiledFiles: {},
        options: options,
        externalsInvoked: false,
        checker: forkChecker ? checker_1.createChecker(compilerInfo, options, webpackOptions) : null,
        cacheIdentifier: cacheIdentifier
    };
}
exports.ensureInstance = ensureInstance;
function setupWatchRun(compiler, instanceName) {
    compiler.plugin('watch-run', function (watching, callback) {
        return __awaiter(this, void 0, Promise, regeneratorRuntime.mark(function _callee2() {
            var compiler, resolver, instance, state, mtimes, changedFiles, tasks;
            return regeneratorRuntime.wrap(function _callee2$(_context2) {
                while (1) {
                    switch (_context2.prev = _context2.next) {
                        case 0:
                            compiler = watching.compiler;
                            resolver = deps_1.createResolver(compiler.options.externals, watching.compiler.resolvers.normal.resolve);
                            instance = resolveInstance(watching.compiler, instanceName);
                            state = instance.tsState;
                            mtimes = watching.compiler.watchFileSystem.watcher.mtimes;
                            changedFiles = Object.keys(mtimes);

                            changedFiles.forEach(function (changedFile) {
                                state.fileAnalyzer.validFiles.markFileInvalid(changedFile);
                            });
                            _context2.prev = 7;
                            tasks = changedFiles.map(function (changedFile) {
                                return __awaiter(this, void 0, Promise, regeneratorRuntime.mark(function _callee() {
                                    return regeneratorRuntime.wrap(function _callee$(_context) {
                                        while (1) {
                                            switch (_context.prev = _context.next) {
                                                case 0:
                                                    if (!/\.ts$|\.d\.ts|\.tsx$/.test(changedFile)) {
                                                        _context.next = 5;
                                                        break;
                                                    }

                                                    _context.next = 3;
                                                    return state.readFileAndUpdate(changedFile);

                                                case 3:
                                                    _context.next = 5;
                                                    return state.fileAnalyzer.checkDependencies(resolver, changedFile);

                                                case 5:
                                                case "end":
                                                    return _context.stop();
                                            }
                                        }
                                    }, _callee, this);
                                }));
                            });
                            _context2.next = 11;
                            return Promise.all(tasks);

                        case 11:
                            state.updateProgram();
                            callback();
                            _context2.next = 19;
                            break;

                        case 15:
                            _context2.prev = 15;
                            _context2.t0 = _context2["catch"](7);

                            console.error(_context2.t0.toString());
                            callback();

                        case 19:
                        case "end":
                            return _context2.stop();
                    }
                }
            }, _callee2, this, [[7, 15]]);
        }));
    });
}
function setupAfterCompile(compiler, instanceName) {
    var forkChecker = arguments.length <= 2 || arguments[2] === undefined ? false : arguments[2];

    compiler.plugin('after-compile', function (compilation, callback) {
        var instance = resolveInstance(compilation.compiler, instanceName);
        var state = instance.tsState;
        if (forkChecker) {
            var payload = {
                files: state.allFiles(),
                resolutionCache: state.host.moduleResolutionHost.resolutionCache
            };
            instance.checker.send({
                messageType: 'compile',
                payload: payload
            });
        } else {
            if (!state.program) {
                state.updateProgram();
            }
            var diagnostics = state.ts.getPreEmitDiagnostics(state.program);
            var emitError = function emitError(err) {
                if (compilation.bail) {
                    console.error('Error in bail mode:', err);
                    process.exit(1);
                }
                compilation.errors.push(new Error(err));
            };
            var errors = helpers_1.formatErrors(instanceName, diagnostics);
            errors.forEach(emitError);
        }
        var phantomImports = [];
        state.allFileNames().forEach(function (fileName) {
            if (!instance.compiledFiles[fileName]) {
                phantomImports.push(fileName);
            }
        });
        instance.compiledFiles = {};
        compilation.fileDependencies.push.apply(compilation.fileDependencies, phantomImports);
        compilation.fileDependencies = _.uniq(compilation.fileDependencies);
        callback();
    });
}
